internal/pipeline/pipeline.go

```diff
@@ -18,99 +18,27 @@ package pipeline
-// source will be the hydrated output of the referenced package.
-type Pipeline struct {
-	yaml.ResourceMeta `yaml:",inline"`
-	//  1. Sources to resolve as input to the pipeline. Possible values:
-	//  a) A slash-separated, OS-agnostic relative package path which may include '.' and '..' e.g. './base', '../foo'
-	//     The source package is resolved recursively.
-	//  b) Resources in this package using '.'. Meta resources such as the Kptfile, Pipeline, and function configs
-	//     are excluded.
-	//  c) Resources in this package AND all resolved subpackages using './*'
-	//
-	// Resultant list of resources are ordered:
-	// - According to the order of sources specified in this array.
-	// - When using './*': Subpackages are resolved in alphanumerical order before package resources.
-	//
-	// When omitted, defaults to './*'.
-	Sources []string `yaml:"sources,omitempty"`
-
-	// 2. Sequence of functions to run. Input of the first function is the resolved sources.
-	// Input of the second function is the output of the first function, and so on.
-	// Order of operations: generators, transformers, validators
-	//
-	// When omitted, defaults to NO-OP.
-	//
-	// 2.a  Sequence of KRM functions that generate resources.
-	Generators []Function `yaml:"generators,omitempty"`
-
-	// 2.b Sequence of KRM functions that transform resources.
-	Transformers []Function `yaml:"transformers,omitempty"`
-
-	// 2.c Sequence of KRM functions that validate resources.
-	// Validators are not permitted to mutate resources.
-	Validators []Function `yaml:"validators,omitempty"`
-}
-
-// String returns the string representation of Pipeline struct
-// The string returned is the struct content in Go default format
-func (p *Pipeline) String() string {
-	return fmt.Sprintf("%+v", *p)
-}
-
-// Validate will validate all fields in the Pipeline
+// ValidatePipeline will validate all fields in the Pipeline
 // 'generators', 'transformers' and 'validators' share same schema and
 // they are valid if all functions in them are ALL valid.
-func (p *Pipeline) Validate() error {
-	if p.APIVersion != kptAPIVersion {
-		return fmt.Errorf("'apiVersion' %q is invalid, should be %q",
-			p.APIVersion, kptAPIVersion)
-	}
-	if p.Kind != pipelineKind {
-		return fmt.Errorf("'kind' %q is invalid, should be %q",
-			p.Kind, pipelineKind)
-	}
-	for i, s := range p.Sources {
-		if s == sourceAllSubPkgs {
-			continue
-		}
-		if err := validatePath(s); err != nil {
-			return fmt.Errorf("'sources[%d]': path %q is invalid: %w", i, s, err)
-		}
-		p.Sources[i] = path.Clean(s)
-	}
-	fns := []Function{}
-	fns = append(fns, p.Generators...)
-	fns = append(fns, p.Transformers...)
+func ValidatePipeline(p *v1alpha2.Pipeline) error {
+	fns := []v1alpha2.Function{}
+	fns = append(fns, p.Mutators...)
 	fns = append(fns, p.Validators...)
-	for _, f := range fns {
-		err := f.Validate()
+	for i := range fns {
+		f := fns[i]
+		err := validateFunction(&f)
```

Comment:
Not a big deal: but you can take struct as input here because validateFunction doesn't mutate the function. While typing this, I realized this can be made receiver and moved in types.go because this is self-contained.

Reply:
I don't think this is real self-contained because it requires some helpers like `validateFunctionName` and `isNodeZero`.

Reply:
moved the validation to v1alpha2 package.